yarpl/include/yarpl/Refcounted.h

```diff
@@ -9,497 +9,87 @@
-}
-
-template <typename T, typename U>
-bool operator<=(const Reference<T>& lhs, const Reference<U>& rhs) noexcept {
-  return lhs.get() <= rhs.get();
+void atomic_store(AtomicReference<T>* ar, Reference<T> r) {
+  std::lock_guard<std::mutex> lg(ar->mtx);
+  ar->ref = std::move(r);
 }
 
+#else
+#ifdef YARPL_WRAP_SHARED_IN_ATOMIC
+// native compiler support for std::atomic<std::shared_ptr>
 template <typename T>
-bool operator<=(const Reference<T>& lhs, std::nullptr_t) noexcept {
-  return lhs.get() <= nullptr;
-}
+using AtomicReference = std::atomic<Reference<T>>;
 
+#else
+// native compiler support for std::atomic_load(std::shared_ptr<...>)
 template <typename T>
-bool operator<=(std::nullptr_t, const Reference<T>& rhs) noexcept {
-  return nullptr <= rhs.get();
-}
-
-template <typename T, typename U>
-bool operator>(const Reference<T>& lhs, const Reference<U>& rhs) noexcept {
-  return lhs.get() > rhs.get();
-}
-
-template <typename T>
-bool operator>(const Reference<T>& lhs, std::nullptr_t) noexcept {
-  return lhs.get() > nullptr;
-}
-
-template <typename T>
-bool operator>(std::nullptr_t, const Reference<T>& rhs) noexcept {
-  return nullptr > rhs.get();
-}
+using AtomicReference = Reference<T>;
+#endif
 
-template <typename T, typename U>
-bool operator>=(const Reference<T>& lhs, const Reference<U>& rhs) noexcept {
-  return lhs.get() >= rhs.get();
+template <typename... T>
+auto atomic_load(T&&... args) {
+  return std::atomic_load(std::forward<T>(args)...);
 }
-
-template <typename T>
-bool operator>=(const Reference<T>& lhs, std::nullptr_t) noexcept {
-  return lhs.get() >= nullptr;
+template <typename... T>
+auto atomic_store(T&&... args) {
+  return std::atomic_store(std::forward<T>(args)...);
 }
-
-template <typename T>
-bool operator>=(std::nullptr_t, const Reference<T>& rhs) noexcept {
-  return nullptr >= rhs.get();
+template <typename... T>
+auto atomic_exchange(T&&... args) {
+  return std::atomic_exchange(std::forward<T>(args)...);
 }
 
-////////////////////////////////////////////////////////////////////////////////
-
-template <typename T, typename CastTo = T, typename... Args>
-Reference<CastTo> make_ref(Args&&... args) {
-  static_assert(
-      std::is_base_of<Refcounted, std::decay_t<T>>::value,
-      "Reference can only be constructed with a Refcounted object");
-
-  static_assert(
-      std::is_base_of<std::decay_t<CastTo>, std::decay_t<T>>::value,
-      "Concrete type must be a subclass of casted-to-type");
-
-  auto r = Reference<CastTo>(
-    new T(std::forward<Args>(args)...),
-    detail::skip_initial_refcount_check{}
-  );
-
-#ifdef YARPL_REFCOUNT_DEBUGGING
-  auto demangled_name = detail::type_name<T>();
-  detail::inc_created(demangled_name);
-  detail::inc_live(demangled_name);
-  detail::set_reference_name{std::move(demangled_name), *r};
 #endif
 
-  return std::move(r);
-}
+class Refcounted {};
 
-class enable_get_ref {
+class enable_get_ref : public std::enable_shared_from_this<enable_get_ref> {
```

Comment:
Why do we need this as opposed to good ol' fashioned `std::enable_shared_from_this<T>`?  Is it because we need _some_ base class to be the shared virtual class?

Reply:
Yep, we just need something to be the inherited class (not even virtual).